Feature: Run grub-mkconfig in background and redirect output to system log

jtomaspm opened the pull request:
This PR modifies the execution of `grub-mkconfig` to run in the background and redirect its output to the system log.
Instead of blocking execution, this allows the command to run asynchronously and still provides a way to view its logs later using `journalctl`.
Doing it this way you don't need to wait for `grub-mkconfig` to finish executing every time we use pacman, while still providing a way to see the output in case something fails.

The hook's full stdout now becomes:
`(5/5) Generating GRUB config to let grub-btrfs detect new snapshots...`

It's still possible to see `grub-mkconfig` logs using:
`$ journalctl -t grub-mkconfig`

Would you be okay with this behavior change, or would you prefer it to be configurable (e.g., an option to choose between synchronous and asynchronous execution)?
I’m happy to adjust the PR if a more flexible or configurable approach is preferred!

Thanks for your time and for maintaining this project 🙏

maximbaz commented:
Hey, thanks for the idea!

A big disclaimer at the beginning, I'm no longer using project myself, so I both won't be able to test and don't have high stakes in the decision.

As a general feedback, my first thought was "wow smart" and my second thought was "this is dangerous"... :sweat_smile: If it is running in background, most people won't be checking logs all that often (I'd maybe check it a couple of times a year at most? :thinking:). If for whatever reason the command failed and your boot is messed up, most likely you will find out the hard way, by not being able to boot next time :sweat_smile: And if to remedy that we start considering sending some sort of notifications on the command failure, I'm wondering if this all is not a big overkill? :thinking: 

But let me know your thoughts on the above! Is it not something you think is a concern, do you see the performance improvement more valuable than this potential downside?

jtomaspm commented:
Thank you for the quick response!

Since you mentioned you’re no longer able to test the script yourself, I’ve attached some screenshots below to showcase the changes.

```
pacman -S vim; date
```
![image](https://github.com/user-attachments/assets/41d441ad-7975-4a03-a95b-304e9871b219)

```
journalctl -t grub-mkconfig
```
![image](https://github.com/user-attachments/assets/f4db412a-65d9-42d5-9f92-9185b27835d3)

As you can see, this change saves the user around 3 seconds each time they use `pacman`.
While it might seem minor, I think this kind of time saving can add up, especially for users who install packages frequently. Depending on the user’s preference, the trade-off could be worth it, even if it means having to check logs manually in case something goes wrong. For others, like people who only do updates from time to time, i understand the time saving is a big unnecessary.
On the other hand, its pretty common to have a high number of snapshots, and as you can see from the journalctl screenshot, that amount of logs appearing each time you use `pacman` can be pretty annoying as well.

I agree that sending a notification on failure would be ideal! The problem with that is that i can't find a way to do it without requiring extra dependencies from the user, like `libnotify`.

I took this idea from the snapper implementation in opensuse, where the logs of the generated config don't show up in the user's console whenever you use the package manager. I have no idea where the logs are being redirected to in their implementation thought.

What do you think about having a synchronous version of `grub-mkconfig` (as is), and another script `grub-mkconfig-async` and let the user choose in the hook file, which script to run.
For compatibility the default would be the already existing `grub-mkconfig`, and I would add some way of letting the user know they can use the async version, for example in a `README.md` file.

maximbaz commented:
Let's do it :)

Do you think it's easier for users to have a separate script file (and modify the hook file), or to keep a single script and choose the logic based on some kind of `SNAP_PAC_GRUB=async` env var that you can set in your shell?

README is a good place to put some additional info about this, please do it.

Thanks!

jtomaspm commented:
Hello 😊

I've updated the script to use the `SNAP_PAC_GRUB_ASYNC` environment variable as the configuration for async mode, as you suggested.

I also added a README file to the project explaining how to configure the hook.
The default behavior hasn't changed, you can now use the `SNAP_PAC_GRUB_ASYNC` variable to run the script asynchronously.

I agree that this is a better approach than maintaining two separate scripts, and it aligns nicely with the existing `SNAP_PAC_SKIP` configuration.

Let me know if there's anything else I should change!

Thanks again!

maximbaz commented:
It's perfect, thanks for the contribution! :partying_face:
